### PR TITLE
Fix Activity http.route and name

### LINKED ISSUE #2967: HTTP OPTIONS request , transaction name(display name) is shown as request path instead of HTTP route ( template ) in asp.net core api app
# Bug Report
HTTP OPTIONS request, transaction name(display name) is shown as **request path** instead of **HTTP route ( template )** in asp.net core api app

Expected transaction name(display name) : **Todo/{tenantid}/{version}**
What is shown is : **/todo/456-123-56abc-x134/2022-02-01**

It seems to happen for HTTP OPTIONS request ,like browser HTTP OPTIONS pre-flight request to check the CORS.

## List of all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):

ASP.NET Core 3.1 , Windows 10
Zipkin & Console exporter is configured 

OTel version
- OpenTelemetry.Exporter.Zipkin1.1.0
- OpenTelemetry.Instrumentation.AspNetCore 1.0.0-rc7
- OpenTelemetry.Extensions.Hosting 1.0.0-rc7
- System.Diagnostics.DiagnosticSource 5.0.1

## Runtime version :

ASP.NET Core 3.1 , Windows 10

## Symptom

For HTTP OPTIONS request,  transaction name(display name) is shown as **request path** instead of **HTTP route ( template )** in asp.net core api app

**What is the expected behavior?**

Expected transaction name (display name):  **Todo/{tenantid}/{version}**

**What is the actual behavior?**

What is shown is : **/todo/456-123-56abc-x134/2022-02-01**

## Reproduce
Any ASP.NET Core 3.1 API app with above packages & send HTTP OPTIONS request ,like browser HTTP OPTIONS pre-flight request to check the CORS.
POST request
```curl
curl -v -X POST http://localhost:5000/todo/456-123-56abc-x134/2022-02-01 -d {}
```
OPTIONS request
```curl
curl -v -X OPTIONS http://localhost:5000/todo/456-123-56abc-x134/2022-02-01 -H "Access-Control-Request-Method: POST"   -H "Access-Control-Request-Headers: content-type"   -H "Origin: https://reqbin.com" -d {}
```
C# Sample
```csharp
 // curl -v -X OPTIONS http://localhost:5000/todo/456-123-56abc-x134/2022-02-01 -H "Access-Control-Request-Method: POST"   -H "Access-Control-Request-Headers: content-type"   -H "Origin: https://reqbin.com" -d {}
    // curl -v -X POST http://localhost:5000/todo/456-123-56abc-x134/2022-02-01 -d {}
    [Route("[controller]")]
    [ApiController]
    public class TodoController : ControllerBase
    {
        static HttpClient httpClient = new HttpClient();
        private readonly ActivitySource activitySource;
        private readonly ILogger<TodoController> _logger;
        public TodoController(ILogger<TodoController> logger,ActivitySource activitySource)
        {
            this.activitySource = activitySource;
            this._logger = logger;
        }
        [HttpPost]
        [Route("{tenantid}/{version}")]
        public IActionResult Create(string tenantid, string version)
        {
            Console.WriteLine($"{nameof(tenantid)}={tenantid},{nameof(version)}={version}");
            using (activitySource.StartActivity("Calling external Service"))
            {
                _logger.LogInformation("TodoController:Get Calling external service");
                InvokeExternalService();
            }
            return Ok();
        }

        public void InvokeExternalService()
        {

            try
            {
                httpClient.GetStringAsync("http://www.google.com").GetAwaiter().GetResult();
            }
            catch (Exception exp)
            {
                // Console.WriteLine(exp);
                throw;
            }
        }
    }
```

We will close this issue if:

* The repro project you share with us is complex. We can't investigate custom
  projects, so don't point us to such, please.
* If we can not reproduce the behavior you're reporting.

## Additional Context
image

image

### LINKED ISSUE #3461: Missing route pattern in requests that do not reach the MVC middleware
# Bug Report

OpenTelemetry NuGet packages:
* OpenTelemetry.Exporter.Console 1.3.0
* OpenTelemetry.Extensions.Hosting 1.0.0-rc9.4
* OpenTelemetry.Instrumentation.AspNetCore 1.0.0-rc9.4

Runtime version:
* `net6.0`

## Symptom
When MVC requests get stopped by previous middlewares (e.g. authentication), the `Microsoft.AspNetCore.Mvc.BeforeAction` event isn't sent and so the `Activity` isn't properly renamed with the route pattern and the `http.route` tag isn't added.

This is a problem since we can't easily identify requests that belong to the same operation due to various failures in middlewares.
In addition, requests that do not use MVC don't have the route set either.

**What is the expected behavior?**

The activity name and `http.route` tag should be equal to the route pattern.

Suggestion for a solution - Instead of relying on `Mvc.BeforeAction` use `Microsoft.AspNetCore.Routing.EndpointMatched`, which would also work outside of MVC. Then get the route pattern using:

```cs
(httpContext.GetEndpoint() as RouteEndpoint)?.RoutePattern?.RawText
```

**What is the actual behavior?**

The activity name equals the full URI path and `http.route` is empty.

## Reproduce

[otel-missing-route.zip](https://github.com/open-telemetry/opentelemetry-dotnet/files/9151522/otel-missing-route.zip)

```cs
public class MyController : Controller
{
    [HttpGet("hasroute/{value}")]
    public int HasRoute(int value) => value;

    [HttpGet("noroute/{value}"), Authorize]
    public int NoRoute(int value) => value;
}
```

Calling those 2 endpoints (while unauthenticated) produces:

```
Activity.DisplayName: hasroute/{value}
http.route: hasroute/{value}
```
and
```
Activity.DisplayName: /noroute/1
```

<details>

```
Activity.TraceId:          0ddb494d4312a1033e8650c55b83b894
Activity.SpanId:           5f78fc1e69504bc6
Activity.TraceFlags:           Recorded
Activity.ActivitySourceName: OpenTelemetry.Instrumentation.AspNetCore
Activity.DisplayName: hasroute/{value}
Activity.Kind:        Server
Activity.StartTime:   2022-07-20T14:58:10.5872211Z
Activity.Duration:    00:00:00.1429528
Activity.Tags:
    http.host: localhost:7196
    http.method: GET
    http.target: /hasroute/1
    http.url: https://localhost:7196/hasroute/1
    http.route: hasroute/{value}
    http.status_code: 200
   StatusCode : UNSET
Resource associated with Activity:
    service.name: unknown_service:WebApplication4

Activity.TraceId:          3d8431517010c153dcb6f681bf50a0a4
Activity.SpanId:           7a30556bc3aa45c3
Activity.TraceFlags:           Recorded
Activity.ActivitySourceName: OpenTelemetry.Instrumentation.AspNetCore
Activity.DisplayName: /noroute/1
Activity.Kind:        Server
Activity.StartTime:   2022-07-20T14:58:15.9094598Z
Activity.Duration:    00:00:00.0347058
Activity.Tags:
    http.host: localhost:7196
    http.method: GET
    http.target: /noroute/1
    http.url: https://localhost:7196/noroute/1
    http.status_code: 401
   StatusCode : UNSET
Resource associated with Activity:
    service.name: unknown_service:WebApplication4
```

</details>

### LINKED ISSUE #3485: ASP.NET Core - Add HTTP method to activity name
# Feature Request

**Is your feature request related to a problem?**

Currently the Activity's display name is set to the MVC template (aka route pattern), which can be confusing as the HTTP method is an integral part of the endpoint's identity.

**Describe the solution you'd like:**

Include the method in the display name, e.g. `PUT keys/{keyid}`

### LINKED ISSUE #3771: [ASP.NET Core] Validate `http.route` value consistency between trace and metric instrumentation
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3755#discussion_r997290570

### PR DESCRIPTION
**Pull request by alanwest**
Fixes #3485
Fixes #3461
Fixes #2967 
Fixes #3771

#### This PR fixes a number of items with respect to the `Activity.DisplayName` and `http.route` attribute:

* `Activity.DisplayName` now conforms to the specification
    * It is set to  `{http.method} {http.route}` when there is a low-cardinality `http.route`
    * It is `{http.method}` when there is not a low-cardinality `http.route`
* `http.route` was missing for some attribute-based routing scenarios
* `http.route` was missing when using minimal API
* `http.route`, when present, is now the same for both the Activity and metric emitted

#### There remain scenarios not fixed by this PR and also changes their behavior. You can review [this readme](https://github.com/open-telemetry/opentelemetry-dotnet/blob/2ce8425902ee64cf555703efd69bb55dbb40cccb/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/README.net8.0.md) for the scenarios that remain unresolved. In summary:

* When using conventional-based routing, `http.route` is set to the matched route template. This does not accurately reflect the actual route when the controller/action invoked are not the defaults. For example:
    * See this [test case](https://github.com/open-telemetry/opentelemetry-dotnet/blob/2ce8425902ee64cf555703efd69bb55dbb40cccb/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/README.net8.0.md#conventionalrouting-non-default-action-with-route-parameter-and-query-string).
        * The path invoked is `/ConventionalRoute/ActionWithStringParameter/2?num=3`
        * `http.route` is set to `{controller=ConventionalRoute}/{action=Default}/{id?}`
        * Ideally, `http.route` would be something like `ConventionalRoute/ActionWithStringParameter/{id?}`.
* When using Razor pages and invoking the root path (e.g. http://example.com/), `http.route` is empty.
    * See this [test case](https://github.com/open-telemetry/opentelemetry-dotnet/blob/2ce8425902ee64cf555703efd69bb55dbb40cccb/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/README.net8.0.md#razorpages-root-path)
        * The path invoked is '/'
        * `http.route` attribute is not set
        * Ideally, `http.route` would be `/Index` because that is the default page.

These scenarios are unfortunate because it limits users from meaningfully querying metrics and faceting them by `http.route`. See #5056 and #5057.

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 1]
```diff
@@ -200,16 +200,12 @@ public void OnEventWritten_New(string name, object payload)
...
-        else
-        {
-            // Set to default "_OTHER" as per spec.
-            // https://github.com/open-telemetry/semantic-conventions/blob/v1.22.0/docs/http/http-spans.md#common-attributes
-            tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpRequestMethod, "_OTHER"));
-        }
+
+        var httpMethod = context.Request.Method;
+        httpMethod = RequestMethodHelper.IsWellKnownHttpMethod(httpMethod)
+            ? httpMethod
```
Comment (JamesNK) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:206]: IsWellKnownHttpMethod ignores case. I think you need to normalize case here. IsWellKnownHttpMethod could return a value that is used.

**Comment by vishweshbankwar** [comment]
Tagging few other issues: #3485 #2967

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/README.net8.0.md [hunk 2]
```diff
@@ -1,37 +1,37 @@
...
+| :green_heart: | AttributeRouting | [Action without parameter](#attributerouting-action-without-parameter) |
+| :green_heart: | AttributeRouting | [Action with parameter](#attributerouting-action-with-parameter) |
+| :green_heart: | AttributeRouting | [Action with parameter before action name in template](#attributerouting-action-with-parameter-before-action-name-in-template) |
+| :green_heart: | AttributeRouting | [Action invoked resulting in 400 Bad Request](#attributerouting-action-invoked-resulting-in-400-bad-request) |
+| :broken_heart: | RazorPages | [Root path](#razorpages-root-path) |
+| :broken_heart: | RazorPages | [Index page](#razorpages-index-page) |
+| :broken_heart: | RazorPages | [Throws exception](#razorpages-throws-exception) |
+| :green_heart: | RazorPages | [Static content](#razorpages-static-content) |
+| :green_heart: | MinimalApi | [Action without parameter](#minimalapi-action-without-parameter) |
+| :green_heart: | MinimalApi | [Action with parameter](#minimalapi-action-with-parameter) |
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/README.net8.0.md:24]: 💚

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 3]
```diff
@@ -41,7 +41,6 @@ internal class HttpInListener : ListenerHandler
     internal const string ActivityOperationName = "Microsoft.AspNetCore.Hosting.HttpRequestIn";
     internal const string OnStartEvent = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start";
     internal const string OnStopEvent = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop";
-    internal const string OnMvcBeforeActionEvent = "Microsoft.AspNetCore.Mvc.BeforeAction";
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:44]: need to remove from [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/ecb5dc8732e8d42d7f3eb4c2b9aa77d4b55a75b8/src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs#L30) as well. That should

**Review by vishweshbankwar — COMMENTED** [review]
Proposed changes looks good to me.

**Review by matt-hensley — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md [hunk 4]
```diff
@@ -41,6 +41,17 @@ exception. The attribute value will be set to full name of exception type.
...
 
+* Fixes the `http.route` attribute for scenarios in which it was
+  previously missing or incorrect. Additionally, the `http.route` attribute
+  is now the same for both the metric and `Activity` emitted for a request.
+  Lastly, the `Activity.DisplayName` has been adjusted to have the format
+  `{http.request.method} {http.route}` to conform with [the specification](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#name).
+  There remain scenarios when using conventional routing or Razor pages where
+  `http.route` is still incorrect. See [#5056](https://github.com/open-telemetry/opentelemetry-dotnet/issues/5056)
+  and [#5057](https://github.com/open-telemetry/opentelemetry-dotnet/issues/5057)
+  for more details.
```
Comment (alanwest) [inline src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:52]: Opened these issues to highlight the remaining known issues related to `http.route`.

@JamesNK I'm interested in your thoughts. I'm hoping we can coordinate to resolve these issues.

**Review by vishweshbankwar — APPROVED** [review]

**Review by utpilla — APPROVED** [review]

**Pull request merged**